File: complex.cpp
#include <stdio.h>
#include <string>

class Complex {
private:
int a, b;

public:
Complex(int real) {
a = real;
b = 0;
}

Complex(int real, int imag) {
a = real;
b = imag;
}

Complex(const Complex &c) {
*this = c;
}

Complex &operator=(const Complex &c) {
a = c.a;
b = c.b;
return *this;
}

Complex &operator+=(const Complex &c) {
a += c.a;
b += c.b;
return *this;
}

Complex &operator-=(const Complex &c) {
a -= c.a;
b -= c.b;
return *this;
}

Complex &operator*=(const Complex &c) {
Complex tmp = *this;
a = tmp.a * c.a - tmp.b * c.b;
b = tmp.a * c.b + tmp.b * c.a;
return *this;
}

std::string str() const {
char buf[32];
if (b == 0) {
snprintf(buf, sizeof(buf), "%d", a);
} else if (a == 0) {
if (b == -1) return "-i";
if (b ==  1) return  "i";
snprintf(buf, sizeof(buf), "%di", b);
} else if (b == -1) {
snprintf(buf, sizeof(buf), "%d-i", a);
} else if (b < 0) {
snprintf(buf, sizeof(buf), "%d%di", a, b);
} else if (b == 1) {
snprintf(buf, sizeof(buf), "%d+i", a);
} else {
snprintf(buf, sizeof(buf), "%d+%di", a, b);
}
return buf;
}
};

Complex operator+(const Complex &c1, const Complex &c2) {
Complex ret = c1;
ret += c2;
return ret;
}

Complex operator-(const Complex &c1, const Complex &c2) {
Complex ret = c1;
ret -= c2;
return ret;
}

Complex operator*(const Complex &c1, const Complex &c2) {
Complex ret = c1;
ret *= c2;
return ret;
}

int main() {
Complex i(0, 1);
Complex a = 1 + 2*i;
Complex b = 3 + 4*i;
Complex c = a + b;
Complex d = a * b;
Complex e = a - 1;
Complex f = a - 2*i;
printf("a = %s\n", a.str().c_str());
printf("b = %s\n", b.str().c_str());
printf("c = %s\n", c.str().c_str());
printf("d = %s\n", d.str().c_str());
printf("e = %s\n", e.str().c_str());
printf("f = %s\n", f.str().c_str());
}

File: complex.scala
object ComplexImplicits {
  implicit def fromInt(re: Int) = new Complex(re, 0)
}

import ComplexImplicits._

case class Complex(real: Int, imag: Int) {
  def this(re: Int) = this(re, 0)

  def +(that: Complex) = Complex(real + that.real, imag + that.imag)
  def +(re: Int) = Complex(real + re, imag)
  def *(that: Complex) = {
    val a = that.real * real - that.imag * imag
    val b = that.real * imag + that.imag * real
    Complex(a, b)
  }
  def *(re: Int) = Complex(real * re, imag)
  def -(that: Complex) = Complex(real - that.real, imag - that.imag)
  def -(re: Int) = Complex(real - re, imag)

  override def toString() = {
    (real, imag) match {
      case (real, imag) if real == 0 && imag == 1 => "i"
      case (real, imag) if real == 0 && imag == -1 => "-i"
      case (real, imag) if real == 0 && imag == 0 => "0"
      case (real, imag) if real == 0 => imag + "i"
      case (real, imag) if imag > 1 => real + " + " + imag + "i"
      case (real, imag) if imag == 1 => real + " + i"
      case (real, imag) if imag == 0 => real.toString
      case (real, imag) if imag == -1 => real + " - i"
      case (real, imag) if imag < -1 => {
        real + " - " + imag.abs + "i"
      }
    }
  }
}

val i = new Complex(0, 1)
val a = 1 + 2 * i
val b = 3 + 4 * i
val c = a + b
val d = a * b
val e = a - 1
val f = a - 2 * i;

println(a)
println(b)
println(c)
println(d)
println(e)
println(f)
